File: testing/core/utils/color_contrast.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

double contrastRatio(Color foreground, Color background) {
  final effectiveForeground = compositeForegroundOverBackground(
    foreground,
    background,
  );
  final lighter = _relativeLuminance(effectiveForeground);
  final darker = _relativeLuminance(background);
  final maxValue = lighter > darker ? lighter : darker;
  final minValue = lighter > darker ? darker : lighter;
  return (maxValue + 0.05) / (minValue + 0.05);
}

Color compositeForegroundOverBackground(Color foreground, Color background) {
  if (foreground.a >= 1) {
    return foreground;
  }
  final alpha = foreground.a.clamp(0, 1).toDouble();
  return Color.fromRGBO(
    _compositeChannel(foreground.r, background.r, alpha),
    _compositeChannel(foreground.g, background.g, alpha),
    _compositeChannel(foreground.b, background.b, alpha),
    1,
  );
}

int _compositeChannel(double foreground, double background, double alpha) {
  final value = (foreground * alpha) + (background * (1 - alpha));
  final normalized = (value * 255).round();
  return normalized.clamp(0, 255);
}

double _relativeLuminance(Color color) {
  double channel(double normalized) {
    if (normalized <= 0.03928) {
      return normalized / 12.92;
    }
    return math.pow((normalized + 0.055) / 1.055, 2.4).toDouble();
  }

  return 0.2126 * channel(color.r) +
      0.7152 * channel(color.g) +
      0.0722 * channel(color.b);
}

File: testing/core/utils/color_contrast.py
from __future__ import annotations

import math

RgbColor = tuple[int, int, int]


def contrast_ratio(foreground: RgbColor, background: RgbColor) -> float:
    lighter = _relative_luminance(foreground)
    darker = _relative_luminance(background)
    maximum = lighter if lighter > darker else darker
    minimum = darker if lighter > darker else lighter
    return (maximum + 0.05) / (minimum + 0.05)


def rgb_to_hex(color: RgbColor) -> str:
    return f"#{color[0]:02X}{color[1]:02X}{color[2]:02X}"


def color_distance(left: RgbColor, right: RgbColor) -> float:
    return math.sqrt(
        ((left[0] - right[0]) ** 2)
        + ((left[1] - right[1]) ** 2)
        + ((left[2] - right[2]) ** 2),
    )


def _relative_luminance(color: RgbColor) -> float:
    def channel(value: int) -> float:
        normalized = value / 255
        if normalized <= 0.03928:
            return normalized / 12.92
        return ((normalized + 0.055) / 1.055) ** 2.4

    return (
        (0.2126 * channel(color[0]))
        + (0.7152 * channel(color[1]))
        + (0.0722 * channel(color[2]))
    )
